### PR TITLE
Allow assigning a readonly static in another instantiation's static ctor for compat

### LINKED ISSUE #990: Breaking change: C# can no longer assign to static readonly field of another instantiation
The following code compiled and ran on previous versions of C#. In Roslyn this does not compile:

```cs
public static class Foo<T>
{
    static Foo()
    {
        if (typeof(T) == typeof(int))
        {
            Foo<int>.compute = x => (int)x;
        }
    }
 
    public static readonly System.Func<double, T> compute;
}
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #990

### CODE REVIEW

**Comment by gafter** [comment]
@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 1]
```diff
@@ -6762,6 +6762,78 @@ static MyDerived()
...
+
+public static class Foo<T>
+{
+    static Foo()
+    {
+        Console.WriteLine(""initializing for "" + typeof(T));
+        Foo<int>.x = typeof(T).Name;
+    }
+
+        public static readonly string x;
```
Comment (VladimirReshetnikov) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs]: Indentation looks wrong.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -1399,7 +1399,9 @@ private bool CheckIsVariable(CSharpSyntaxNode node, BoundExpression expr, BindVa
                     if ((object)containing != null &&
                         fieldIsStatic == containing.IsStatic &&
                         (fieldIsStatic || fieldAccess.ReceiverOpt.Kind == BoundKind.ThisReference) &&
-                        fieldSymbol.ContainingType == containing.ContainingType)
+                        (Compilation.FeatureStrictEnabled
```
Comment (VladimirReshetnikov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: A comment would be really useful here.

**Comment by VladimirReshetnikov** [comment]
:+1:

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 3]
```diff
@@ -6762,6 +6762,77 @@ static MyDerived()
...
+    }
+
+    public static readonly int X;
+    public static int Y { get; }
+}";
+            CreateCompilationWithMscorlib(text, options: TestOptions.ReleaseDll).VerifyDiagnostics(
+                // (6,9): error CS0200: Property or indexer 'Foo<int>.Y' cannot be assigned to -- it is read only
+                //         Foo<int>.Y = 2;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyProp, "Foo<int>.Y").WithArguments("Foo<int>.Y").WithLocation(6, 9)
+                );
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs]: Should we also confirm that the values were assigned as expected?

**Comment by AlekseyTs** [comment]
LGTM

**Comment by AlekseyTs** [comment]
Does Roslyn have the same behavior change in VB?

**Comment by gafter** [comment]
@AlekseyTs Yes; see #1028. However, given the slightly greater VB tolerance for breaking incorrect code, that may be the right place to land.

**Pull request merged**